src/core/ext/filters/client_channel/lb_policy/child_policy_handler.cc

```diff
@@ -0,0 +1,277 @@
+//
+// Copyright 2018 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/ext/filters/client_channel/lb_policy/child_policy_handler.h"
+
+#include "absl/strings/str_cat.h"
+
+#include "src/core/ext/filters/client_channel/lb_policy_registry.h"
+
+namespace grpc_core {
+
+//
+// ChildPolicyHandler::Helper
+//
+
+class ChildPolicyHandler::Helper
+    : public LoadBalancingPolicy::ChannelControlHelper {
+ public:
+  explicit Helper(RefCountedPtr<ChildPolicyHandler> parent)
+      : parent_(std::move(parent)) {}
+
+  ~Helper() { parent_.reset(DEBUG_LOCATION, "Helper"); }
+
+  RefCountedPtr<SubchannelInterface> CreateSubchannel(
+      const grpc_channel_args& args) override {
+    if (parent_->shutting_down_) return nullptr;
+    if (!CalledByCurrentChild() && !CalledByPendingChild()) return nullptr;
+    return parent_->channel_control_helper()->CreateSubchannel(args);
+  }
+
+  void UpdateState(grpc_connectivity_state state,
+                   std::unique_ptr<SubchannelPicker> picker) override {
+    if (parent_->shutting_down_) return;
+    // If this request is from the pending child policy, ignore it until
+    // it reports READY, at which point we swap it into place.
+    if (CalledByPendingChild()) {
+      if (GRPC_TRACE_FLAG_ENABLED(*(parent_->tracer_))) {
+        gpr_log(GPR_INFO,
+                "[child_policy_handler %p] helper %p: pending child policy %p "
+                "reports state=%s",
+                parent_.get(), this, child_, ConnectivityStateName(state));
+      }
+      if (state != GRPC_CHANNEL_READY) return;
+      grpc_pollset_set_del_pollset_set(
+          parent_->child_policy_->interested_parties(),
+          parent_->interested_parties());
+      parent_->child_policy_ = std::move(parent_->pending_child_policy_);
+    } else if (!CalledByCurrentChild()) {
```

Comment:
Can you give an example of what you mean by "a positive condition"?  I'm not sure what you mean.

Note that I can't just change this to check that it *is* from the current child and then move line 67 inside of the body here, because we want line 67 to run even if the case of a pending child that has just gone READY and been swapped into being the current child.

Reply:
> Note that I can't just change this to check that it is from the current child

That was what I meant. And I see the point here; you are right.